### PR TITLE
Smoke test for the Oracle WebLogic 

### PR DESCRIPTION
**Pull request by vovencij**
Smoke test is built using same concepts as existing smoke tests in this repo and in `opentelemetry-java-instrumentation`. It covers officially supported WebLogic versions - WLS 12 on Java 8 and WLS 14 on Java 8 and 11.

Because of yet unclear (work in progress) legal arrangements for building and publishing shared images with WebLogic (which could be used in GitHub actions and environments alike), test is expecting local WebLogic images. If there are no images, then test is ignored. To build local development images of Oracle WebLogic one has to follow the instructions in `src/weblogic/README.md` and download some files from Oracle, while agreeing to its development license terms. 

Because the smoke test for WLS has one more dimension in addition to just `jdk`, `SmokeTest` and `SpringBootTest` were updated to provide a supplier function for the image name lookup instead of an abstract method taking `jdk` as an input argument. This part sort of redoes PR https://github.com/signalfx/splunk-otel-java/pull/25 while solving the issue https://github.com/signalfx/splunk-otel-java/issues/20

image

### CODE REVIEW

File: smoke-tests/src/test/java/com/splunk/opentelemetry/SpringBootSmokeTest.java [hunk 1]
```diff
@@ -16,43 +16,47 @@
...
 class SpringBootSmokeTest extends SmokeTest {
 
-  protected String getTargetImage(int jdk) {
-    return "open-telemetry-docker-dev.bintray.io/java/smoke-springboot-jdk" + jdk + ":latest";
+  private static Stream<Arguments> springBootConfigurations() {
+    return Stream.of(
+        arguments(
+            new SpringBootConfiguration(8),
+            new SpringBootConfiguration(11),
+            new SpringBootConfiguration(14),
```
Comment (iNikem) [inline smoke-tests/src/test/java/com/splunk/opentelemetry/SpringBootSmokeTest.java:41]: No need for 14. Rule of thumb is "all LTS plus the latest if not LTS"

File: smoke-tests/src/test/java/com/splunk/opentelemetry/WebLogicSmokeTest.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    String currentAgentVersion = getCurrentAgentVersion();
+
+    Response response = client.newCall(request).execute();
+    Collection<ExportTraceServiceRequest> traces = waitForTraces();
+
+    String responseBody = response.body().string();
+    Assertions.assertTrue(
+        responseBody.contains("bytes read by weblogic.net.http.SOAPHttpURLConnection"));
+    // TODO: When https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1439
```
Comment (iNikem) [inline smoke-tests/src/test/java/com/splunk/opentelemetry/WebLogicSmokeTest.java:80]: It is merged

File: smoke-tests/src/weblogic/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
+# Demo application running in Oracle WebLogic
+
+## Prerequisites
+To package the demo into WebLogic you need Docker _development_ images from the official Oracle GitHub repo:
+https://github.com/oracle/docker-images
+To run build the images locally you will need to follow instructions (this means that certain zip files with
```
Comment (iNikem) [inline smoke-tests/src/weblogic/README.md:6]:
```suggestion
To build the images locally you will need to follow instructions (this means that certain zip files with
```

File: smoke-tests/src/weblogic/src/main/docker/wls/container-scripts/create-wls-domain.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import os
```
Comment (iNikem) [inline smoke-tests/src/weblogic/src/main/docker/wls/container-scripts/create-wls-domain.py:1]: All these files in this directory, where are they from? Who is the original author and what is their license?

File: smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java [hunk 5]
```diff
@@ -93,9 +95,12 @@ static void setupSpec() {
 
   protected GenericContainer target;
 
-  void startTarget(int jdk) {
+  /** @param targetImageNameSupplier function that returns an image name to start. */
+  void startTarget(Supplier<String> targetImageNameSupplier) {
```
Comment (iNikem) [inline smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java:99]: Why do you want `Supplier` here? And not just simple `String`?

File: smoke-tests/src/test/java/com/splunk/opentelemetry/SpringBootSmokeTest.java (hunk 1 above)
Comment (vovencij) [reply smoke-tests/src/test/java/com/splunk/opentelemetry/SpringBootSmokeTest.java:41]: I just converted what there was already: https://github.com/signalfx/splunk-otel-java/blob/main/smoke-tests/src/test/java/com/splunk/opentelemetry/SpringBootSmokeTest.java#L44

But will remove.

File: smoke-tests/src/test/java/com/splunk/opentelemetry/WebLogicSmokeTest.java (hunk 2 above)
Comment (vovencij) [reply smoke-tests/src/test/java/com/splunk/opentelemetry/WebLogicSmokeTest.java:80]: Correct me if I'm wrong, but that fix is not yet released, and splunk distro depends on the released version of openetelemtry-java-instrumentation, so this fix is not yet available here?

File: smoke-tests/src/weblogic/src/main/docker/wls/container-scripts/create-wls-domain.py (hunk 4 above)
Comment (vovencij) [reply smoke-tests/src/weblogic/src/main/docker/wls/container-scripts/create-wls-domain.py:1]: I wrote them based on examples in https://github.com/oracle/docker-images/tree/master/OracleWebLogic/dockerfiles/12.2.1.4/container-scripts and other relevant subprojects.

File: smoke-tests/src/test/java/com/splunk/opentelemetry/SpringBootSmokeTest.java (hunk 1 above)
Comment (iNikem) [inline smoke-tests/src/test/java/com/splunk/opentelemetry/SpringBootSmokeTest.java:41]:
```suggestion
```

File: smoke-tests/src/test/java/com/splunk/opentelemetry/WebLogicSmokeTest.java (hunk 2 above)
Comment (iNikem) [reply smoke-tests/src/test/java/com/splunk/opentelemetry/WebLogicSmokeTest.java:80]: that's true

File: smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java (hunk 5 above)
Comment (vovencij) [reply smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java]: `String` is good. Supplier was a remnant of some intermediary refactoring step. Replaced.

**Comment by vovencij** [comment]
Spring boot smoke tests are currently failing because of this: https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1546

File: smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java [hunk 6]
```diff
@@ -93,9 +94,11 @@ static void setupSpec() {
 
   protected GenericContainer target;
 
-  void startTarget(int jdk) {
+  void startTarget(String targetImageName) {
     target =
-        new GenericContainer<>(getTargetImage(jdk))
+        new GenericContainer<>(targetImageName)
+            .waitingFor(Wait.forListeningPort())
```
Comment (iNikem) [inline smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java:100]: Isn't it default?

File: smoke-tests/src/test/java/com/splunk/opentelemetry/WebLogicSmokeTest.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+/**
+ * This is the smoke test for OpenTelemetry Java agent with Oracle WebLogic. The test is ignored if
+ * there are no WebLogic images installed locally. See the manual in src/weblogic directory for
+ * instructions on how to build required images.
+ */
+class WebLogicSmokeTest extends SmokeTest {
+
+  @Override
+  protected Map<String, String> getExtraEnv() {
+    return Collections.singletonMap("SPLUNK_OTEL_CONFIG_SPAN_PROCESSOR_INSTRLIB_ENABLED", "true");
```
Comment (iNikem) [inline smoke-tests/src/test/java/com/splunk/opentelemetry/WebLogicSmokeTest.java:50]: You can remove this now, see #34

File: smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java [hunk 8]
```diff
@@ -57,14 +60,12 @@
...
 
-  protected abstract String getTargetImage(int jdk);
-
   /** Subclasses can override this method to customise target application's environment */
   protected Map<String, String> getExtraEnv() {
     return Collections.emptyMap();
   }
 
-  private static GenericContainer backend;
+  protected static GenericContainer backend;
```
Comment (iNikem) [inline smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java:68]: It can be private, right?

**Review by iNikem — APPROVED** [review]

File: smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java (hunk 6 above)
Comment (vovencij) [reply smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java]: Yes, removed.

File: smoke-tests/src/test/java/com/splunk/opentelemetry/WebLogicSmokeTest.java (hunk 7 above)
Comment (vovencij) [reply smoke-tests/src/test/java/com/splunk/opentelemetry/WebLogicSmokeTest.java]: Removed.

File: smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java (hunk 8 above)
Comment (vovencij) [reply smoke-tests/src/test/java/com/splunk/opentelemetry/SmokeTest.java]: yes.

**Pull request merged**